commons/src/main/java/org/infinispan/commons/collections/MiniSet.java

```diff
@@ -0,0 +1,320 @@
+package org.infinispan.commons.collections;
+
+import java.io.Serializable;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * Set optimized for few keys. The optimization is mostly memory-wise, reducing amount of objects created.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class MiniSet<T> extends AbstractSet<T> implements Serializable {
+    private static final Object DUMMY = new Object();
+    private static final int ARRAY_SIZE = 8;
+    public static final int MAX_OPTIMIZED_CAPACITY = ARRAY_SIZE;
+
+    /*
+     union { T, T[], Map<T, Object> }
+     */
+    private Object holder;
+    protected Type type = Type.EMPTY;
+
+    public MiniSet() {}
+
+    public MiniSet(T... entries) {
+        if (entries.length == 0) {
+            // nothing to do
+        } else if (entries.length == 1) {
+            holder = entries[0];
+            if (holder == null) throw new IllegalArgumentException();
+            type = Type.SINGLE;
+        } else if (entries.length < ARRAY_SIZE) {
+            Object[] array = new Object[ARRAY_SIZE];
```

Comment:
why not using add()? like

```
holder = new Object[ARRAY_SIZE];
type = Type.ARRAY;
for (T entry : entries) add(entry);
```

the code will be simplified and the performance would be the same. WDYT?